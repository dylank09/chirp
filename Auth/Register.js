import React, { useState } from "react";
import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  KeyboardAvoidingView,
} from "react-native";

import { theme } from "../assets/Theme";
import textLogo from "../assets/logo-text.png";
import ChirpButton from "../components/ChirpButton";
import AuthTextBox from "../components/AuthTextBox";
import AuthAlert from "../components/AuthAlert";
import AuthProviderButton from "../components/AuthProviderButton";

import firebase from "../config/FirebaseConfig";
import "firebase/auth";
import GoogleSignIn from "./GoogleSignIn";
import AddUserToDB from "../userFunctions/AddUserToDB";

export default function Register({ navigation }) {
  const [fullName, setFullName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [nameError, setNameError] = useState("");
  const [emailError, setEmailError] = useState("");
  const [passwordError, setPasswordError] = useState("");

  const next = "ChirpGroups";

  function createUserEmailPass() {
    if (validateName() && validatePassword()) {
      firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then((userCredential) => {
          setEmailError("");
          // Signed in
          // var user = userCredential.user;
          AddUserToDB(fullName);
          navigation.navigate(next);
          // set currentUser in the App to some value in order to render the main screen?
        })
        .catch((error) => {
          console.log(error);
          switch (error.code) {
            case "auth/invalid-email":
              setEmailError("Email address is not valid.");
              break;
            case "auth/blablabla":
              setEmailError("User already exists.");
              break;
            default:
              setEmailError("Incorrect details. Try again.");
          }
        });
    }
  }

  function validatePassword() {
    var check = false;
    if (password.length < 8) {
      setPasswordError("Password is too short.");
    } else if (!(/[A-Z]/.test(password) && /[0-9]/.test(password))) {
      setPasswordError("Password must include a capital and a number.");
    } else if (password != confirmPassword) {
      setPasswordError("Passwords don't match.");
    } else {
      setPasswordError("");
      check = true;
    }
    return check;
  }

  function validateName() {
    if (fullName.length < 2) {
      setNameError("Name is too short.");
      return false;
    } else {
      setNameError("");
      return true;
    }
  }

  function navToLogin() {
    navigation.navigate("Login");
  }

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.headingText}>Register</Text>
      <KeyboardAvoidingView style={styles.inputContainer} behavior="padding">
        <AuthTextBox
          placeholder="Full name"
          onChangeText={setFullName}
        ></AuthTextBox>
        <AuthAlert text={nameError}></AuthAlert>
        <AuthTextBox placeholder="Email" onChangeText={setEmail}></AuthTextBox>
        <AuthAlert text={emailError}></AuthAlert>
        <AuthTextBox
          placeholder="Password"
          onChangeText={setPassword}
        ></AuthTextBox>
        <AuthAlert text={passwordError}></AuthAlert>
        <AuthTextBox
          placeholder="Confirm password"
          onChangeText={setConfirmPassword}
        ></AuthTextBox>
      </KeyboardAvoidingView>
      <ChirpButton
        style={styles.continueButton}
        text="Continue"
        onPress={createUserEmailPass}
      ></ChirpButton>
      <Text style={styles.orText}>or</Text>
      <View style={styles.line} />
      <AuthProviderButton
        text="Sign in with Google"
        type="google"
        onPress={() => GoogleSignIn(navigation, next)}
      ></AuthProviderButton>
      <AuthProviderButton
        text="Sign in with Microsoft"
        type="microsoft"
        onPress={() => GoogleSignIn(navigation, next)}
      ></AuthProviderButton>
      <Text style={styles.loginText} onPress={navToLogin}>
        Already have an account? Sign in
      </Text>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
    backgroundColor: theme.colors.background,
    alignItems: "center",
    justifyContent: "space-between",
  },
  inputContainer: {
    flex: 2,
    width: "100%",
    alignItems: "center",
    justifyContent: "center",
    marginBottom: 20,
  },
  headingText: {
    flex: 0.5,
    fontSize: 24,
    color: theme.colors.text,
    marginTop: 15,
    marginBottom: 10,
  },
  orText: {
    fontSize: theme.dimensions.standardFontSize,
    color: theme.colors.text,
  },
  line: {
    borderBottomColor: theme.colors.text,
    borderBottomWidth: 1,
    marginBottom: 7,
    width: "50%",
  },
  loginText: {
    flex: 0.3,
    color: theme.colors.placeholderColor,
    fontSize: theme.dimensions.standardFontSize,
    marginTop: 8,
    marginBottom: 5,
    textDecorationLine: "underline",
  },
});
