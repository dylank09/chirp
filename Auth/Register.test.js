import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react-native";

import Register from "./Register";

describe("<Register />", () => {
  const tree = renderer.create(<Register />).toJSON();

  it("renders correctly", () => {
    expect(tree).toMatchSnapshot();
  });

  it("continue register button is present", () => {
    const { getByTestId } = render(<Register />);
    const foundButton = getByTestId("registerButton");
    expect(foundButton).toBeTruthy();
  });

  it("Google button is present", () => {
    const { getByTestId } = render(<Register />);
    const foundButton = getByTestId("googleButton");
    expect(foundButton).toBeTruthy();
  });

  it("navigate to login button is present", () => {
    const { getByTestId } = render(<Register />);
    const foundButton = getByTestId("navToLoginButton");
    expect(foundButton).toBeTruthy();
  });
});
