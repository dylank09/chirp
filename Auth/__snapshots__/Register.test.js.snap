// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(76, 76, 71)",
      "flex": 1,
      "justifyContent": "space-around",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 24,
        "marginBottom": 10,
        "marginTop": 15,
      }
    }
  >
    Register
  </Text>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginBottom": 20,
          "width": "100%",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={true}
      multiline={false}
      onChangeText={[Function]}
      placeholder="Full name"
      placeholderTextColor="rgb(158, 158, 158)"
      rejectResponderTermination={true}
      secureTextEntry={false}
      selectTextOnFocus={false}
      selectionColor="rgb(193, 73, 83)"
      style={
        Array [
          Object {
            "borderBottomColor": "rgb(255, 255, 255)",
            "borderBottomWidth": 1,
            "color": "rgb(255, 255, 255)",
            "width": "82%",
          },
          Object {
            "fontSize": 15,
            "marginBottom": 4,
            "marginTop": 7,
            "padding": 5,
          },
        ]
      }
      testID="chirpTextBox"
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "color": "rgb(193, 73, 83)",
          "fontSize": 12,
          "height": 15,
          "margin": 0,
          "paddingLeft": 3,
          "width": "75%",
        }
      }
    >
       
    </Text>
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={true}
      multiline={false}
      onChangeText={[Function]}
      placeholder="Email"
      placeholderTextColor="rgb(158, 158, 158)"
      rejectResponderTermination={true}
      secureTextEntry={false}
      selectTextOnFocus={false}
      selectionColor="rgb(193, 73, 83)"
      style={
        Array [
          Object {
            "borderBottomColor": "rgb(255, 255, 255)",
            "borderBottomWidth": 1,
            "color": "rgb(255, 255, 255)",
            "width": "82%",
          },
          Object {
            "fontSize": 15,
            "marginBottom": 4,
            "marginTop": 7,
            "padding": 5,
          },
        ]
      }
      testID="chirpTextBox"
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "color": "rgb(193, 73, 83)",
          "fontSize": 12,
          "height": 15,
          "margin": 0,
          "paddingLeft": 3,
          "width": "75%",
        }
      }
    >
       
    </Text>
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={true}
      multiline={false}
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="rgb(158, 158, 158)"
      rejectResponderTermination={true}
      secureTextEntry={true}
      selectTextOnFocus={true}
      selectionColor="rgb(193, 73, 83)"
      style={
        Array [
          Object {
            "borderBottomColor": "rgb(255, 255, 255)",
            "borderBottomWidth": 1,
            "color": "rgb(255, 255, 255)",
            "width": "82%",
          },
          Object {
            "fontSize": 15,
            "marginBottom": 4,
            "marginTop": 7,
            "padding": 5,
          },
        ]
      }
      testID="chirpTextBox"
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "color": "rgb(193, 73, 83)",
          "fontSize": 12,
          "height": 15,
          "margin": 0,
          "paddingLeft": 3,
          "width": "75%",
        }
      }
    >
       
    </Text>
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={true}
      multiline={false}
      onChangeText={[Function]}
      placeholder="Confirm password"
      placeholderTextColor="rgb(158, 158, 158)"
      rejectResponderTermination={true}
      secureTextEntry={true}
      selectTextOnFocus={true}
      selectionColor="rgb(193, 73, 83)"
      style={
        Array [
          Object {
            "borderBottomColor": "rgb(255, 255, 255)",
            "borderBottomWidth": 1,
            "color": "rgb(255, 255, 255)",
            "width": "82%",
          },
          Object {
            "fontSize": 15,
            "marginBottom": 4,
            "marginTop": 7,
            "padding": 5,
          },
        ]
      }
      testID="chirpTextBox"
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(193, 73, 83)",
        "borderRadius": 18,
        "height": undefined,
        "margin": 10,
        "opacity": 1,
        "padding": 10,
        "width": "40%",
      }
    }
    testID="registerButton"
  >
    <Text
      style={
        Object {
          "color": "rgb(255, 255, 255)",
          "fontSize": 15,
        }
      }
    >
      Continue
       
    </Text>
  </View>
  <Text
    style={
      Object {
        "borderBottomColor": "rgb(255, 255, 255)",
        "borderBottomWidth": 1,
        "color": "rgb(255, 255, 255)",
        "fontSize": 15,
        "textAlign": "center",
        "width": "40%",
      }
    }
  >
    or
  </Text>
  <View />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 18,
        "flexDirection": "row",
        "justifyContent": "center",
        "margin": 8,
        "opacity": 1,
        "paddingBottom": 7,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 7,
        "width": "70%",
      }
    }
    testID="googleButton"
  >
    <Image
      source={1}
      style={
        Object {
          "height": 19,
          "margin": 3,
          "width": 19,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "black",
          "fontWeight": "600",
          "marginLeft": 5,
        }
      }
    >
      Sign in with Google
       
    </Text>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "rgb(158, 158, 158)",
        "fontSize": 15,
        "marginBottom": 5,
        "marginTop": 8,
        "textDecorationLine": "underline",
      }
    }
    testID="navToLoginButton"
  >
    Already have an account? Sign in
  </Text>
</RCTSafeAreaView>
`;
